firmware/config/boards/proteus/board.mk

```diff
@@ -25,3 +25,6 @@ endif
 
 # Override DEFAULT_ENGINE_TYPE
 DDEFS += $(MCU_DEFS) -DEFI_USE_OSC=TRUE -DEFI_FATAL_ERROR_PIN=GPIOE_3 -DFIRMWARE_ID=\"proteus\" -DDEFAULT_ENGINE_TYPE=PROTEUS -DUSE_ADC3_VBATT_HACK -DSTM32_ADC_USE_ADC3=TRUE -DEFI_INCLUDE_ENGINE_PRESETS=FALSE
+
+#Enable on-chip flash
+DDEFS += -DEFI_INTERNAL_FLASH=TRUE
```

Comment:
why is this enabled for proteus?

Reply:
oh, wait, this is internal flash.  Should internal flash be the default?

Reply:
It was always enabled. efifeatures.h in proteus directory includes "../../stm32f4ems/efifeatures.h" where it was enabled.
Now when we have two types of flash. And two drivers for internal flash (ChibiOS backported and RusEFI own)
I see no other way to enable SPI flash for MRE while keep it disabled for all other board that re-use stm32f4ems/efifeatures.h.
So I disable both flash options in common stm32f4ems/efifeatures.h. And enable it for each board in its board.mk.
Not sure if it is correct way. But now there is total mess with configuration files. (efifeatures.h halconf.h and board.mk).

Reply:
Yeah, I agree that it's gotten a bit out of control.  I was just wondering why not have a single define that toggles it, instead of a define to enable each.

Reply:
There is another issue with RusEFI driver vs ChibiOS driver. RusEFI's supports flash for all SMT32 and Kinetic that were ever used in RusEFI HW.
While ChibiOS supports only 407 and 437.
So we have 4 options:
1. internal flash with RusEFI driver (used if EFI_INTERNAL_FLASH=true, EFI_FLASH_SPI=false and HAL_USE_FLASH=false)
2. internal flash with ChibiOS driver (used if EFI_INTERNAL_FLASH=true, EFI_FLASH_SPI=false and HAL_USE_FLASH=true)
3. SPI flash with ChibiOS driver (used if EFI_FLASH_SPI=true and HAL_USE_FLASH=true,  EFI_INTERNAL_FLASH don't care)
4. no flash at all ( EFI_INTERNAL_FLASH=false, EFI_FLASH_SPI=false)
I was not able to find better way than introduce EFI_FLASH_SPI.
BTW, common settings read/write code in enable by EFI_CONFIGURATION_STORAGE (= (EFI_FLASH_SPI || EFI_INTERNAL_FLASH)

Reply:
can we improve ChibiOS to add support for F7 flash?

Reply:
iirc it's the same, it just needs a `__DSB` before and after the actual flash copy step

Reply:
yep, we're running the same implementation on both, with this extra on the f7:
```
	// Cortex-M7 (STM32F7/H7) can execute out order - need to force a full flush
	// so that we actually wait for the operation to complete!
#if CORTEX_MODEL == 7
	__DSB();
#endif
```